pytorch_lightning/plugins/collective/torch_collective.py

```diff
@@ -0,0 +1,113 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Optional, Union
+
+import torch
+import torch.distributed
+
+from pytorch_lightning.overrides.torch_distributed import broadcast_object_list
+from pytorch_lightning.plugins.collective import Collective
+from pytorch_lightning.utilities import _TORCH_GREATER_EQUAL_1_8
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.distributed import all_gather_ddp_if_available, distributed_available
+from pytorch_lightning.utilities.distributed import group as _group
+from pytorch_lightning.utilities.distributed import ReduceOp, sync_ddp_if_available
+from pytorch_lightning.utilities.types import _METRIC_COLLECTION
+
+
+class TorchCollective(Collective):
+    """Collective interface for DDP, DDPSpawn, DP and DDP2."""
+
+    def __init__(
+        self,
+        local_reduce: bool = False,
```

Comment:
I don't believe this should be here. I think `TorchCollective` should only be used for encapsulating the difference between different distributed backends. The `local_reduce` functionality is not related to `torch.distributed` and should be handled by upper layers.

Reply:
For DP and DDP collective have different behaviors, should we have separate collective for each type then?

Reply:
IMHO since local reduce is not a behavior offered by the distributed backends we are trying to encapsulate, the logic should just live where it's used.

Reply:
I see, you mean pass it in to specific functions like reduce(), yeah make sense, will do!

Reply:
+1 , unclear whats local_reduce :)

Reply:
It's the logic in current dp and ddp2

Reply:
> I see, you mean pass it in to specific functions like reduce(), yeah make sense, will do!

No I actually meant the logic for handling local reduce should be moved out of `Collective`. Because (1) it's not a functionality provided by the underlying distributed backend (2) it's not an operation that all members of the process group participate (3) it's not supported in other implementations

Reply:
Discussed and resolved offline. Due to current lightning structure, we will keep as is and add more comments